File: backend/app/api/v1/projects.py
from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.orm import Session
from uuid import UUID
from typing import Optional

from app.dependencies.database import get_db
from app.dependencies.auth import get_current_user_dep as get_current_user
from app.models.user import User
from app.schemas.project import (
    ProjectCreate,
    ProjectUpdate,
    ProjectResponse,
    ProjectListResponse,
)
from app.schemas.responses import SuccessResponse
from app.services.project_service import ProjectService

router = APIRouter()


@router.get(
    "",
    response_model=SuccessResponse[ProjectListResponse],
)
def get_projects(
    page: int = Query(1, ge=1),
    size: int = Query(20, ge=1, le=100),
    search: Optional[str] = None,
    language: Optional[str] = None,
    favorite: Optional[bool] = None,
    sort_by: str = Query(
        "updated_at",
        pattern="^(updated_at|created_at|last_opened_at|name)$",
    ),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ProjectService(db)

    result = service.get_projects(
        user_id=current_user.id,
        page=page,
        size=size,
        search=search,
        language=language,
        favorite=favorite,
        sort_by=sort_by,
        deleted=False,
    )

    return SuccessResponse(
        message="Projects retrieved successfully.",
        data=result,
    )


@router.get(
    "/trash",
    response_model=SuccessResponse[ProjectListResponse],
)
def get_trashed_projects(
    page: int = Query(1, ge=1),
    size: int = Query(20, ge=1, le=100),
    search: Optional[str] = None,
    sort_by: str = Query(
        "updated_at",
        pattern="^(updated_at|created_at|name)$",
    ),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ProjectService(db)

    result = service.get_projects(
        user_id=current_user.id,
        page=page,
        size=size,
        search=search,
        sort_by=sort_by,
        deleted=True,
    )

    return SuccessResponse(
        message="Trash retrieved successfully.",
        data=result,
    )


@router.post(
    "",
    response_model=SuccessResponse[ProjectResponse],
    status_code=status.HTTP_201_CREATED,
)
def create_project(
    data: ProjectCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ProjectService(db)

    project = service.create_project(data, current_user.id)

    return SuccessResponse(
        message="Project created successfully.",
        data=project,
    )


@router.get(
    "/{project_id}",
    response_model=SuccessResponse[ProjectResponse],
)
def get_project(
    project_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ProjectService(db)

    project = service.get_project(project_id, current_user.id)

    return SuccessResponse(
        message="Project retrieved successfully.",
        data=project,
    )


@router.put(
    "/{project_id}",
    response_model=SuccessResponse[ProjectResponse],
)
def update_project(
    project_id: UUID,
    data: ProjectUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ProjectService(db)

    project = service.update_project(project_id, data, current_user.id)

    return SuccessResponse(
        message="Project updated successfully.",
        data=project,
    )


@router.delete(
    "/trash",
    response_model=SuccessResponse[dict],
)
def empty_trash(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ProjectService(db)
    service.empty_trash(current_user.id)
    return SuccessResponse(
        message="Trash emptied successfully.",
        data={},
    )


@router.delete(
    "/{project_id}",
    response_model=SuccessResponse[ProjectResponse],
)
def soft_delete_project(
    project_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ProjectService(db)

    project = service.soft_delete_project(project_id, current_user.id)

    return SuccessResponse(
        message="Project moved to trash successfully.",
        data=project,
    )

@router.delete(
    "/{project_id}/permanent",
    response_model=SuccessResponse[dict],
)
def hard_delete_project(
    project_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ProjectService(db)

    service.hard_delete_project(project_id, current_user.id)

    return SuccessResponse(
        message="Project permanently deleted.",
        data={},
    )


@router.post(
    "/{project_id}/restore",
    response_model=SuccessResponse[ProjectResponse],
)
def restore_project(
    project_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ProjectService(db)

    project = service.restore_project(project_id, current_user.id)

    return SuccessResponse(
        message="Project restored successfully.",
        data=project,
    )


@router.post(
    "/{project_id}/favorite",
    response_model=SuccessResponse[ProjectResponse],
)
def toggle_favorite(
    project_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ProjectService(db)

    project = service.toggle_favorite(project_id, current_user.id)

    return SuccessResponse(
        message="Favorite updated successfully.",
        data=project,
    )


@router.post(
    "/{project_id}/duplicate",
    response_model=SuccessResponse[ProjectResponse],
    status_code=status.HTTP_201_CREATED,
)
def duplicate_project(
    project_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    service = ProjectService(db)

    project = service.duplicate_project(project_id, current_user.id)

    return SuccessResponse(
        message="Project duplicated successfully.",
        data=project,
    )

File: src/lib/api/projects.ts
import { fetchApi } from "../api";

export type ProjectVisibility = "PRIVATE" | "UNLISTED" | "PUBLIC";

export interface Project {
  id: string;
  owner_id: string;
  name: string;
  description: string | null;
  language: string | null;
  visibility: ProjectVisibility;
  favorite: boolean;
  color: string | null;
  icon: string | null;
  slug: string;
  created_at: string;
  updated_at: string;
  last_opened_at: string;
  deleted_at: string | null;
}

export interface ProjectListResponse {
  items: Project[];
  total: number;
  page: number;
  size: number;
}

export interface CreateProjectPayload {
  name: string;
  description?: string;
  language?: string;
  visibility?: ProjectVisibility;
  color?: string;
  icon?: string;
}

export interface UpdateProjectPayload extends Partial<CreateProjectPayload> {
  favorite?: boolean;
}

export const projectsApi = {
  getProjects: async (params?: {
    page?: number;
    size?: number;
    search?: string;
    language?: string;
    favorite?: boolean;
    sort_by?: "updated_at" | "created_at" | "last_opened_at" | "name";
  }) => {
    const query = new URLSearchParams();
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined) query.append(key, value.toString());
      });
    }
    return fetchApi(`/projects?${query.toString()}`);
  },

  getTrash: async (params?: {
    page?: number;
    size?: number;
    search?: string;
    sort_by?: "updated_at" | "created_at" | "name";
  }) => {
    const query = new URLSearchParams();
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined) query.append(key, value.toString());
      });
    }
    return fetchApi(`/projects/trash?${query.toString()}`);
  },

  getProject: async (id: string) => {
    return fetchApi(`/projects/${id}`);
  },

  createProject: async (data: CreateProjectPayload) => {
    return fetchApi("/projects", {
      method: "POST",
      body: JSON.stringify(data),
    });
  },

  updateProject: async (id: string, data: UpdateProjectPayload) => {
    return fetchApi(`/projects/${id}`, {
      method: "PUT",
      body: JSON.stringify(data),
    });
  },

  deleteProject: async (id: string) => {
    return fetchApi(`/projects/${id}`, {
      method: "DELETE",
    });
  },

  permanentDeleteProject: async (id: string) => {
    return fetchApi(`/projects/${id}/permanent`, {
      method: "DELETE",
    });
  },

  emptyTrash: async () => {
    return fetchApi("/projects/trash", {
      method: "DELETE",
    });
  },

  restoreProject: async (id: string) => {
    return fetchApi(`/projects/${id}/restore`, {
      method: "POST",
    });
  },

  toggleFavorite: async (id: string) => {
    return fetchApi(`/projects/${id}/favorite`, {
      method: "POST",
    });
  },

  duplicateProject: async (id: string) => {
    return fetchApi(`/projects/${id}/duplicate`, {
      method: "POST",
    });
  },
};
